test/load-generator/acme/directory.go

```diff
@@ -0,0 +1,208 @@
+func (e ErrInvalidEndpointURL) Error() string {
+	return fmt.Sprintf(
+		"directoryURL JSON had invalid URL value (%q) for %q endpoint",
+		e.value, e.endpoint)
+}
+
+// Directory is a type for holding URLs extracted from the ACME server's
+// Directory resource.
+//
+// See RFC 8555 Section 7.1.1 "Directory".
+//
+// Its public API is read-only and therefore it is safe for concurrent access.
+type Directory struct {
+	// endpointURLs is a map from endpoint name to
+	endpointURLs map[Endpoint]*url.URL
+}
+
+// getRawDirectory validates the provided directoryURL and makes a GET request
+// to fetch the raw bytes of the server's directory resource. If the URL is
+// invalid, if there is an error getting the directory bytes, or if the HTTP
+// response code is not 200 an error is returned.
+func getRawDirectory(directoryURL string) ([]byte, error) {
+	if directoryURL == "" {
+		return nil, ErrEmptyDirectory
+	}
+
+	if _, err := url.Parse(directoryURL); err != nil {
+		return nil, ErrInvalidDirectoryURL
+	}
+
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout:   10 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			TLSHandshakeTimeout: 5 * time.Second,
+			TLSClientConfig: &tls.Config{
+				// Bypassing CDN or testing against Pebble instances can cause
+				// validation failures. For a **test-only** tool its acceptable to skip
+				// cert verification of the ACME server's HTTPs certificate.
+				InsecureSkipVerify: true,
+			},
+			MaxIdleConns:    1,
+			IdleConnTimeout: 15 * time.Second,
+		},
+		Timeout: 10 * time.Second,
+	}
+
+	resp, err := httpClient.Get(directoryURL)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrInvalidDirectoryHTTPCode
+	}
+
+	rawDirectory, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return rawDirectory, nil
+}
+
+// NewDirectory creates a Directory populated from the ACME directory resource
+// returned by a GET request to the provided directoryURL. An error is returned
+// if the directory URL is invalid, or if there is a problem connecting to the
+// directory URL, unmarshaling the JSON directory resource, or if the directory
+// is missing required endpoints (see RequiredEndpoints).
+func NewDirectory(directoryURL string) (*Directory, error) {
+	// Fetch the raw directory JSON
+	dirContents, err := getRawDirectory(directoryURL)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the directory
+	var dirResource map[string]interface{}
+	if err := json.Unmarshal(dirContents, &dirResource); err != nil {
+		return nil, ErrInvalidDirectoryJSON
+	}
+
+	// serverURL tries to find a valid url.URL for the provided endpoint in
+	// the unmarshaled directory resource.
+	serverURL := func(name Endpoint) (*url.URL, error) {
+		if rawURL, ok := dirResource[string(name)]; !ok {
```

Comment:
Oh boy, this block is, uh, intense.

Reply:
Is there something you want changed?

Reply:
Eh, I can't really think of a better way to structure it, it's just a bit painful to read.